EIPS/eip-2997.md

```diff
@@ -0,0 +1,100 @@
+
+## Specification
+
+`IMPERSONATECALL`: `0xf6`, takes 7 operands:
+
+- `gas`: the amount of gas the code may use in order to execute;
+- `to`: the destination address whose code is to be executed;
+- `in_offset`: the offset into memory of the input;
+- `in_size`: the size of the input in bytes;
+- `ret_offset`: the offset into memory of the output;
+- `ret_size`: the size of the scratch pad for the output.
+- `salt` is a `32` bytes value (a stack item). 
+
+### Computation of impersonated sender
+
+The impersonated sender address is computed as `keccak256( 0xff ++ address ++ salt ++ zeros32)[12:]`.
+
+- `0xff` is a single byte, 
+- `address` is always `20` bytes, and represents the address of the real caller contract.
+- `salt` is always `32` bytes. 
+
+- The field zeros32 corresponds to 32 zero bytes.
+
+This scheme emulates `CREATE2` address derivation, but it cannot practically collude with the `CREATE2` address space.
+
+### Notes
+- The opcode behaves exactly as `CALL` in terms of gas consumption.
+- In the called context `CALLER (0x33)` returns the impersonated address.
+- If value transfer is non-zero in the call, the value is transferred from the impersonated account, and not from the real caller. This can be used to transfer ether out of an impersonated account.
+
+
+
+## Rationale
+
+Even if `IMPERSONATECALL` requires hashing 3 words, implying an additional cost of 180 gas, we think the benefit of accounting for hashing doesn't not compensate increasing the complexity of the implementation.
+
+We use the zeros32 field to base address derivation in a pre-image of similar size than CREATE2 and reuse the existing address derivation functions. We also avoid worrying about address collisions between EOA derivation (65 bytes pre-image), CREATE derivation (from 23 to 27 bytes pre-image, for a 32bit nonce) and CREATE2 derivation (85 bytes pre-image). 
+
+An option is to omit the zeros32 field: the resulting length of the Keccak pre-image for IMPERSONATECALL address is 53 bytes , which does not generate address collision.
+
+While the same functionality could be provided in a pre-compiled contract, we believe using a new opcode is a cleaner solution.
+
+
+## Possible arguments against
+
+* You can replicate this functionality with counterfactual contract creation. We argue that the there is an important benefit of avoiding the deployment cost in case the sponsor needs to serve thousands of users.
+
```

Comment:
EIPs are not the appropriate place to make arguments/pleas.  Consider moving this content to the motivation section and re-wording it as a motivation.